src/values.rs

```diff
@@ -3,213 +3,198 @@
 }
 
-impl<Elem: ValueOfType> ContainerValue<Elem> {
-    pub(crate) fn desc(&self) -> String {
+impl Value {
+    /// Returns the name of this [`Value`].
+    pub fn name(&self) -> String {
         match self {
-            ContainerValue::Sequence(vals) => {
-                let names: Vec<_> = vals.iter().map(ValueOfType::name).collect();
+            Value::Prim(p) => p.name(),
+            Value::Tuple(vals) => {
+                let names: Vec<_> = vals.iter().map(Value::name).collect();
                 format!("const:seq:{{{}}}", names.join(", "))
             }
-            ContainerValue::Sum(tag, val) => format!("const:sum:{{tag:{tag}, val:{}}}", val.name()),
+            Value::Sum(tag, val) => format!("const:sum:{{tag:{tag}, val:{}}}", val.name()),
         }
     }
-    pub(crate) fn check_container(&self, ty: &Container<Elem::T>) -> Result<(), ConstTypeError> {
-        match (self, ty) {
-            (ContainerValue::Sequence(elems), Container::Tuple(tup_tys)) => {
-                if elems.len() != tup_tys.len() {
-                    return Err(ConstTypeError::TupleWrongLength);
-                }
-                for (elem, ty) in elems.iter().zip(tup_tys.iter()) {
-                    elem.check_type(ty)?;
-                }
-                Ok(())
-            }
-            (ContainerValue::Sequence(elems), Container::Array(elem_ty, sz)) => {
-                if elems.len() != *sz {
-                    return Err(ConstTypeError::TupleWrongLength);
-                }
-                for elem in elems {
-                    elem.check_type(elem_ty)?;
-                }
-                Ok(())
-            }
-            (ContainerValue::Sum(tag, value), Container::Sum(variants)) => {
-                value.check_type(variants.get(*tag).ok_or(ConstTypeError::InvalidSumTag)?)
-            }
-            (_, Container::Alias(s)) => Err(ConstTypeError::NoAliases(s.to_string())),
-            (_, _) => Err(ValueOfType::container_error(ty.clone(), self.clone())),
-        }
+
+    /// Description of the value.
+    pub fn description(&self) -> &str {
+        "Constant value"
     }
 
-    pub(crate) fn map_vals<T2: ValueOfType>(&self, f: &impl Fn(Elem) -> T2) -> ContainerValue<T2> {
-        match self {
-            ContainerValue::Sequence(vals) => {
-                ContainerValue::Sequence(vals.iter().cloned().map(f).collect())
-            }
-            ContainerValue::Sum(tag, value) => {
-                ContainerValue::Sum(*tag, Box::new(f((**value).clone())))
-            }
-        }
+    /// Constant unit type (empty Tuple).
+    pub const fn unit() -> Self {
+        Self::Tuple(vec![])
+    }
+
+    /// Constant Sum over units, used as predicates.
+    pub fn simple_predicate(tag: usize) -> Self {
+        Self::sum(tag, Self::unit())
+    }
+
+    /// Constant Sum over Tuples with just one variant of unit type
+    pub fn simple_unary_predicate() -> Self {
+        Self::simple_predicate(0)
```

Comment:
```suggestion
        Self::simple_predicate(1)
```
?

Reply:
don't think so - this is a value, and the tag is zero indexed.